### PR TITLE
add auto pod upgrade controller for daemoset

### PR DESCRIPTION
**Pull request by xavier-hou**
Signed-off-by: hxcGit <houxc_mail@163.com>

#### What this PR does / why we need it:
#921 

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

**Comment by rambohe-ch** [comment]
@hxcGit please fix the errors of github actions.

**Comment by rambohe-ch** [comment]
@hxcGit we should consider the name of controller carefully. and i think `podupgrade` for newly added controller is not a good name.

**Comment by hxcGit** [comment]
> @hxcGit we should consider the name of controller carefully. and i think `podupgrade` for newly added controller is not a good name.

Yes, it's just a temporary name.

File: pkg/controller/podupgrade/pod_upgrade_controller.go [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+		daemonsetSynced: daemonsetInformer.Informer().HasSynced,
+
+		nodeLister: nodeInformer.Lister(),
+		nodeSynced: nodeInformer.Informer().HasSynced,
+
+		podLister: podInformer.Lister(),
+		podSynced: podInformer.Informer().HasSynced,
+
+		daemonsetWorkqueue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		nodeWorkqueue:      workqueue.New(),
```
Comment (rambohe-ch) [inline pkg/controller/podupgrade/pod_upgrade_controller.go:70]: please merge `nodeWorkqueue` and `daemonsetWorkQueue` into one workqueue.

File: pkg/controller/podupgrade/pod_upgrade_controller.go [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+		},
+	})
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(old, new interface{}) {
+			oldNode := old.(*corev1.Node)
+			newNode := new.(*corev1.Node)
+			if !NodeReady(&oldNode.Status) && NodeReady(&newNode.Status) {
+				klog.Infof("node %q turn to ready", newNode.Name)
+				ctrl.nodeWorkqueue.Add(newNode.Name)
```
Comment (rambohe-ch) [inline pkg/controller/podupgrade/pod_upgrade_controller.go:98]: please calculate the corresponding OTA and auto daemonsets, and queue these daemonsets into daemonset workqueue.

File: pkg/controller/podupgrade/pod_upgrade_controller.go [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+			}
+		},
+	})
+	return &ctrl
+}
+
+func (c *Controller) enqueueDaemonset(obj interface{}) {
```
Comment (rambohe-ch) [inline pkg/controller/podupgrade/pod_upgrade_controller.go:105]: line 105~111 maybe should be placed at Line88, it's not need to new func.

File: pkg/controller/podupgrade/pod_upgrade_controller.go [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+		klog.Error("sync podupgrade controller timeout")
+	}
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runDaemonsetWorker, time.Second, stopCh)
+	}
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runNodeWorker, time.Second, stopCh)
+	}
```
Comment (rambohe-ch) [inline pkg/controller/podupgrade/pod_upgrade_controller.go:137]: please remove node worker.

File: pkg/controller/podupgrade/pod_upgrade_controller.go [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+			// Only control daemonset with annotation "apps.openyurt.io/upgrade-strategy"
+			if !metav1.HasAnnotation(newDS.ObjectMeta, UpgradeAnnotation) || !metav1.HasAnnotation(oldDS.ObjectMeta, UpgradeAnnotation) {
+				return
+			}
+
+			// TODO: change to compare generation and revision hash
+			if newDS.ResourceVersion == oldDS.ResourceVersion || reflect.DeepEqual(newDS.Spec.Template.Spec, oldDS.Spec.Template.Spec) {
+				return
+			}
+
```
Comment (rambohe-ch) [inline pkg/controller/podupgrade/pod_upgrade_controller.go:87]: we also need to make sure the native upgrade strategy should be `OnDelete`.

File: pkg/controller/podupgrade/util.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+	for i, pod := range pods {
+		if pod.Spec.NodeName == node.Name {
+			nodePods = append(nodePods, pods[i])
+			nodePodsNames = append(nodePodsNames, pod.Name)
+		}
+	}
+
+	if len(nodePodsNames) > 0 {
+		klog.V(5).Infof("Daemonset %v has pods %v", node.Name, nodePodsNames)
```
Comment (rambohe-ch) [inline pkg/controller/podupgrade/util.go:99]: Daemonset --> Node

**Comment by rambohe-ch** [comment]
@hxcGit when auto upgrade mode is used, we need to consider the number of unavailable pods should be under some limit, so only limited pods will be crash when end user make a mistake to upgrade(like use a out-of-use image).

**Comment by hxcGit** [comment]
> @hxcGit when auto upgrade mode is used, we need to consider the number of unavailable pods should be under some limit, so only limited pods will be crash when end user make a mistake to upgrade(like use a out-of-use image).

@rambohe-ch I think kubernetes rollingUpdate offers a nice implementation. How about reusing its source code? The modification required maybe is to judge whether nodes are ready when calculating the number of unavailable pods.
For pods on not-ready nodes do not count in the total number of unavailable pods.

**Comment by rambohe-ch** [comment]
> > @hxcGit when auto upgrade mode is used, we need to consider the number of unavailable pods should be under some limit, so only limited pods will be crash when end user make a mistake to upgrade(like use a out-of-use image).
> 
> @rambohe-ch I think kubernetes rollingUpdate offers a nice implementation. How about reusing its source code? The modification required maybe is to judge whether nodes are ready when calculating the number of unavailable pods. For pods on not-ready nodes do not count in the total number of unavailable pods.

@hxcGit I agree with you. go ahead~

File: cmd/yurt-controller-manager/app/controllermanager.go [hunk 7]
```diff
@@ -305,7 +305,7 @@ func NewControllerInitializers() map[string]InitFunc {
 	controllers := map[string]InitFunc{}
 	controllers["nodelifecycle"] = startNodeLifecycleController
 	controllers["yurtcsrapprover"] = startYurtCSRApproverController
-
+	controllers["podupgrade"] = startPodUpgradeController
```
Comment (rambohe-ch) [inline cmd/yurt-controller-manager/app/controllermanager.go:308]: how about rename `podupgrade` to `daemonpodupdater`?

File: pkg/controller/podupdater/pod_updater_controller.go [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+// This annotation will only be added if the upgrade strategy is "apps.openyurt.io/upgrade-strategy":"ota".
+const PodUpgradableAnnotation = "apps.openyurt.io/pod-upgradable"
+
+const MaxUnavailableAnnotation = "apps.openyurt.io/max-unavailable"
+
+const (
+	// BurstReplicas is a rate limiter for booting pods on a lot of pods.
+	// The value of 250 is chosen b/c values that are too high can cause registry DoS issues.
+	BurstReplicas = 250
+)
```
Comment (rambohe-ch) [inline pkg/controller/podupdater/pod_updater_controller.go:68]: please merge line 48~68 into one `const`

File: pkg/controller/podupdater/pod_updater_controller.go [hunk 9]
```diff
@@ -0,0 +1,502 @@
+/*
```
Comment (rambohe-ch) [inline pkg/controller/podupdater/pod_updater_controller.go:1]: how about rename file name `pod_updater_controller.go` to `daemon_pod_updater_controller.go`?

File: cmd/yurt-controller-manager/app/controllermanager.go (hunk 7 above)
Comment (hxcGit) [reply cmd/yurt-controller-manager/app/controllermanager.go:308]: OK. I'm trying separate this controller into a base controller and a daemonset controller as your advice. I'll modify the name then.

File: pkg/controller/podupdater/kubernetes/controller_util.go [hunk 10]
```diff
@@ -0,0 +1,646 @@
+/*
```
Comment (rambohe-ch) [inline pkg/controller/podupdater/kubernetes/controller_util.go:1]: please add unit tests for this file.

File: pkg/controller/podupdater/kubernetes/daemonset_util.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
+/*
```
Comment (rambohe-ch) [inline pkg/controller/podupdater/kubernetes/daemonset_util.go:1]: ditto

File: pkg/controller/podupdater/kubernetes/pod_util.go [hunk 12]
```diff
@@ -0,0 +1,83 @@
+/*
```
Comment (rambohe-ch) [inline pkg/controller/podupdater/kubernetes/pod_util.go:1]: ditto

File: pkg/controller/podupdater/pod_updater_controller.go (hunk 8 above)
Comment (hxcGit) [reply pkg/controller/podupdater/pod_updater_controller.go:68]: I separate these constant for different usage before. I'll merge them.

File: pkg/controller/podupdater/util.go [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+// SetPodUpgradeAnnotation calculate and set annotation "apps.openyurt.io/pod-upgradable" to pod
+func SetPodUpgradeAnnotation(clientset client.Interface, ds *appsv1.DaemonSet, pod *corev1.Pod) error {
+	ok := IsDaemonsetPodLatest(ds, pod)
+
+	var res bool
+	if !ok {
+		res = true
+	}
+
+	metav1.SetMetaDataAnnotation(&pod.ObjectMeta, PodUpgradableAnnotation, strconv.FormatBool(res))
```
Comment (rambohe-ch) [inline pkg/controller/podupdater/util.go:122]: please delete line 117~120, and rename `ok` to `isUpgraded`, then modify this line as following:
> metav1.SetMetaDataAnnotation(&pod.ObjectMeta, PodUpgradableAnnotation, strconv.FormatBool(!isUpgraded))

File: pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go [hunk 14]
```diff
@@ -0,0 +1,507 @@
...
+		// Check if node is ready, ignore not-ready node
+		ready, err := NodeReadyByName(c.nodeLister, nodeName)
+		if err != nil {
+			return fmt.Errorf("couldn't check node %q ready status, %v", nodeName, err)
+		}
+		if !ready {
+			continue
+		}
+		for _, pod := range pods {
+			if err := SetPodUpdateAnnotation(c.kubeclientset, ds, pod); err != nil {
```
Comment (rambohe-ch) [inline pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go:307]: it's not a good idea to a new annotation to specify that pod should be upgraded. I think we can add a new pod status condition for pod to specify pod upgrade, and the info as following:
```
condition type: PodNeedUpdate
condition status: True/False
```

Comment (xavier-hou) [reply pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go]: `PodNeedUpgrade` or `PodNeedUpdate` ?
I found I got `PodNeedUpgrade` in github comment while `PodNeedUpdate` in email notification.

Comment (rambohe-ch) [reply pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go]: > `PodNeedUpgrade` or `PodNeedUpdate` ? I found I got `PodNeedUpgrade` in github comment while `PodNeedUpdate` in email notification.

I have modified `PodNeedUpdate` to `PodNeedUpgrade`. so please use `PodNeedUpgrade`.

File: pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go [hunk 15]
```diff
@@ -0,0 +1,527 @@
+/*
+Copyright 2022 The OpenYurt Authors.
+
```
Comment (zyjhtangtang) [inline pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go:3]: need to add copyright  “Copyright 2017 The Kubernetes Authors.”

File: pkg/controller/daemonpodupdater/util.go [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+	return nil
+}
+
+// checkPrerequisites checks that daemonset meets two conditions
+// 1. annotation "apps.openyurt.io/update-strategy"="auto" or "ota"
+// 2. update strategy is "OnDelete"
+func checkPrerequisites(ds *appsv1.DaemonSet) bool {
+	v, ok := ds.Annotations[UpdateAnnotation]
+	if !ok || (v != "auto" && v != "ota") {
+		return false
```
Comment (zyjhtangtang) [inline pkg/controller/daemonpodupdater/util.go:147]: Update strategy "auto" and "ota" using already defined constants.

File: pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go [hunk 17]
```diff
@@ -0,0 +1,527 @@
...
+// while not, set pod condition "PodNeedUpgrade" to "true"
+func (c *Controller) otaUpdate(ds *appsv1.DaemonSet, nodeToDaemonPods map[string][]*corev1.Pod) error {
+	for nodeName, pods := range nodeToDaemonPods {
+		// Check if node is ready, ignore not-ready node
+		ready, err := NodeReadyByName(c.nodeLister, nodeName)
+		if err != nil {
+			return fmt.Errorf("couldn't check node %q ready status, %v", nodeName, err)
+		}
+		if !ready {
+			continue
```
Comment (zyjhtangtang) [inline pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go:322]: In the OTA update strategy, need to set the update status for not ready nodes.

File: pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go [hunk 18]
```diff
@@ -0,0 +1,527 @@
...
+				// The old pod isn't available, so it needs to be replaced
+				klog.V(5).Infof("DaemonSet %s/%s pod %s on node %s is out of date and not available, allowing replacement", ds.Namespace, ds.Name, oldPod.Name, nodeName)
+				// Record the replacement
+				if allowedReplacementPods == nil {
+					allowedReplacementPods = make([]string, 0, len(nodeToDaemonPods))
+				}
+				allowedReplacementPods = append(allowedReplacementPods, oldPod.Name)
+			case numUnavailable >= maxUnavailable:
+				// No point considering any other candidates
+				continue
```
Comment (zyjhtangtang) [inline pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go:387]: Can you return directly and wait for next sync?

File: pkg/controller/daemonpodupdater/daemon_pod_updater_controller_test.go [hunk 19]
```diff
@@ -0,0 +1,513 @@
+/*
+Copyright 2022 The OpenYurt Authors.
+
```
Comment (zyjhtangtang) [inline pkg/controller/daemonpodupdater/daemon_pod_updater_controller_test.go:3]: need to add copyright “Copyright 2017 The Kubernetes Authors.”

File: pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go (hunk 18 above)
Comment (xavier-hou) [reply pkg/controller/daemonpodupdater/daemon_pod_updater_controller.go:387]: Referring to the K8s `daemonset controller`, old unavailable pods do not count in the maxUnavailable number, so do not return directly here may be better.

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/update.go#L69

**Review by rambohe-ch — APPROVED** [review]

**Pull request merged**